model/incoming_webhook.go

```diff
@@ -5,29 +5,59 @@ package model
 
 import (
 	"bytes"
+	"crypto/hmac"
+	"crypto/sha1"
+	"crypto/sha256"
+	"encoding/hex"
 	"encoding/json"
+	"hash"
 	"io"
 	"net/http"
 	"regexp"
+	"strconv"
+	"strings"
 )
 
 const (
 	DEFAULT_WEBHOOK_USERNAME = "webhook"
+	WHITE_IP_LIST_SIZE       = 5 //upper limit
 )
 
+type SignedIncomingHook struct {
```

Comment:
Do all of these need to be exported in the public `model` or can we keep them unexported in the `app` package?

Reply:
Right. These are only for web API which receives webhook requests from clients so I moved to web package, private.